File: 47. Permutations II.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public void backtrack(int[] nums, StringBuilder currStr, List<Integer> currArr, HashSet<Integer> takenInd, HashSet<String> addedPerms,int n){
        boolean lastEle = true;
        for(int i = 0; i < n; i++){
            if(!takenInd.contains(i)){
                lastEle = false;
                currStr.append("#").append(Integer.toString(nums[i]));
                currArr.add(nums[i]);
                takenInd.add(i);
                backtrack(nums, currStr, currArr, takenInd, addedPerms, n);
                while(currStr.charAt(currStr.length() - 1) != '#') currStr.deleteCharAt(currStr.length() - 1);
                currStr.deleteCharAt(currStr.length() - 1);
                takenInd.remove(i);
                currArr.remove(currArr.size() - 1);
            }
        }

        if(lastEle){
            if(!addedPerms.contains(currStr.toString())){
                ans.add(new ArrayList<>(currArr));
                addedPerms.add(currStr.toString());
                // System.out.println(currStr.toString());
            }
        }
    }
    public List<List<Integer>> permuteUnique(int[] nums) {
        HashSet<Integer> takenInd = new HashSet<>();
        HashSet<String> addedPerms = new HashSet<>();
        int n = nums.length;
        for(int i = 0; i < n; i++){
            takenInd.add(i);
            List<Integer> currArr = new ArrayList<>();
            currArr.add(nums[i]);
            backtrack(nums, new StringBuilder(Integer.toString(nums[i])), currArr, takenInd, addedPerms, n);
            takenInd.remove(i);
        }

        return ans;
    }
}

File: 47. Permutations II.cpp
class Solution {
public:
    vector<vector<int>> ans;

    void backtrack(vector<int>& nums, string currStr, vector<int>& currArr,
                   unordered_set<int>& takenInd, unordered_set<string>& addedPerms, int n) {

        bool lastEle = true;
        for (int i = 0; i < n; i++) {
            if (takenInd.find(i) == takenInd.end()) {
                lastEle = false;
                string nextStr = currStr + "#" + to_string(nums[i]);
                currArr.push_back(nums[i]);
                takenInd.insert(i);

                backtrack(nums, nextStr, currArr, takenInd, addedPerms, n);

                takenInd.erase(i);
                currArr.pop_back();
            }
        }

        if (lastEle) {
            if (addedPerms.find(currStr) == addedPerms.end()) {
                ans.push_back(currArr);
                addedPerms.insert(currStr);
            }
        }
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        unordered_set<int> takenInd;
        unordered_set<string> addedPerms;
        int n = nums.size();

        for (int i = 0; i < n; i++) {
            takenInd.insert(i);
            vector<int> currArr = {nums[i]};
            string currStr = to_string(nums[i]);
            backtrack(nums, currStr, currArr, takenInd, addedPerms, n);
            takenInd.erase(i);
        }

        return ans;
    }
};